moshi/src/main/java/com/squareup/moshi/Types.java

```diff
@@ -284,11 +284,21 @@ public static Type collectionElementType(Type context, Class<?> contextRawType)
     Type mapType = getSupertype(context, contextRawType, Map.class);
     if (mapType instanceof ParameterizedType) {
       ParameterizedType mapParameterizedType = (ParameterizedType) mapType;
-      return mapParameterizedType.getActualTypeArguments();
+      return getCanonicalizedMapTypeArguments(mapParameterizedType);
     }
     return new Type[] { Object.class, Object.class };
   }
 
+  private static Type[] getCanonicalizedMapTypeArguments(ParameterizedType parameterizedType) {
+    Type[] types = parameterizedType.getActualTypeArguments();
+    for (int i = 0; i < types.length; i++) {
+      if (types[i] instanceof GenericArrayType) {
+        types[i] = getRawType(types[i]);
```

Comment:
sigh, this isn’t strictly correct. Imagine `Map<String, List<String>[]>`. The value’s raw type is `List<?>[]`, but we want to retain `List<String>[]`. It’s so absurdly shitty that Java has two ways of modeling `byte[]`.